File: haskell/TruthTableGenerator.hs
module TruthTableGenerator where 

    import Lit


    --                                  0 Funcoes para obter as proposicoes da tabela

    {-
    Verifica se um char NAO estah contido em uma string.
    Ex:
    notIn 'a' "abcd" = False
    notIn 'e' "abcd" = True
    -}
    notIn :: Char -> [Char] -> Bool
    notIn x [] = True
    notIn x (head:tail) | x == head = False
                        | otherwise = notIn x tail

    {-
    Remove char's repetidos de uma string. (Usa uma lista auxiliar inicialmente vazia).
    Ex:
    removeRepetitions [] "pppq" = "pq"
    -}
    removeRepetitions :: [Char] -> [Char] -> [Char]
    removeRepetitions list [] = []
    removeRepetitions list (head:tail) | notIn head list = [head] ++ removeRepetitions (list ++ [head]) tail
                                       | otherwise = removeRepetitions list tail

    {-
    Obtem as proposicoes recebendo a expressao, mas com repeticoes.
    Ex:
    getPropositionsBrute "(p | (p & q))" = "ppq"
    -}
    getPropositionsBrute :: [Char] -> [Char]
    getPropositionsBrute [] = []
    getPropositionsBrute (head:tail) | notIn head "|&()*#~ " = [head] ++ getPropositionsBrute tail
                                     | otherwise = getPropositionsBrute tail

    {-
    Obtem as proposicoes recebendo a expressao.
    Ex:
    getPropositionsBrute "(p | (p & q))" = "pq"
    getPropositionsBrute "(p | (p & r))" = "pqr"
    -}
    getPropositions :: [Char] -> [Char]
    getPropositions expression = removeRepetitions [] (getPropositionsBrute expression)


    --                                   Funcoes para gerar a tabela de possibilidades

    {-
    Gera uma sequência de n zeros.
    Ex:
    zeroSequence 5 = "00000"
    -}
    zeroSequence :: Int -> [Char]
    zeroSequence 0 = ""
    zeroSequence n = "0" ++ zeroSequence (n-1)

    {-
    Completa um binário incompleto com zeros à esquerda para que o número de caracteres seja controlado.
    Ex:
    completeBinary 5 "101" = "00101"
    -}
    completeBinary :: Int -> [Char] -> [Char]
    completeBinary n bin = zeroSequence (n - (length bin)) ++ bin

    {-
    Transforma um inteiro em um binario.
    Ex:
    getBinaryIncomplete 5 = "101"
    -}
    getBinaryIncomplete :: Int -> [Char]
    getBinaryIncomplete 0 = ""
    getBinaryIncomplete n | n `mod` 2 == 1 = getBinaryIncomplete (quot n 2) ++ "1"
                          | n `mod` 2 == 0 = getBinaryIncomplete (quot n 2) ++ "0"

    {-
    Transforma um inteiro em um binario, mas com controle no tamanho de caracteres.
    Ex:
    getBinary 5 6 = "000101"
    -}
    getBinary :: Int -> Int -> [Char]
    getBinary 0 size = completeBinary size "0"
    getBinary n size = completeBinary size (getBinaryIncomplete n)

    {-
    Retorna uma tabela com todas as interpretacoes possiveis para um determinado numero de variaveis.
    Ex:
    getPossibilitiesTable 2 3 = ["00","01","10","11"]
    -}
    getPossibilitiesTable :: Int -> Int -> [[Char]]
    getPossibilitiesTable size (-1) = []
    getPossibilitiesTable size n = getPossibilitiesTable size (n-1) ++ [getBinary n size]

    {-
    Faz o calculo para saber o numero de linhas da tabela e chama o metodo para obtê-la
    Ex:
    callGetPossibilitiesTable 2 = ["00","01","10","11"]
    -}
    callGetPossibilitiesTable :: Int -> [[Char]]
    callGetPossibilitiesTable n = getPossibilitiesTable n ((2 ^ n) - 1)


    --                                   Funcoes para obter os resultados

    {-
    Obtem a posicao de uma proposicao na string de proposicoes.
    Ex:
    getPropPosition "p" "pqr" = 0
    getPropPosition "q" "pqr" = 1
    -}
    getPropPosition :: [Char] -> [Char] -> Int
    getPropPosition prop (h:t) | prop == [h] = 0
                               | otherwise = 1 + getPropPosition prop t 

    {-
    Obtem o valor de uma proposicao de acordo com sua posicao em relacao a interpretacao.
    Ex:
    getPropositionValue "p" "pqr" "010" = False
    getPropositionValue "q" "pqr" "010" = True
    -}
    getPropositionValue :: [Char] -> [Char] -> [Char] -> Bool
    getPropositionValue prop list poss | (poss !! (getPropPosition prop list)) == '0' = False
                                       | otherwise = True

    {-
    Inverte o valor de uma proposicao baseado no seu operador unario.
    Ex:
    checkUnOP True "~" = False
    checkUnOP True "" = True
    -}
    checkUnOP :: Bool -> String -> Bool
    checkUnOP value unOp | unOp == "~" = not value
                         | otherwise = value

    {-
    Varre toda a arvore do literal para obter o valor da expressao, realizando as operacoes da arvore para determinada interpretacao.
    Ex:
    literalToString <literal> = "(a & b)"
    result <literal> "ab" "00" = False
    result <literal> "ab" "11" = True
    -}
    result :: Literal -> [Char] -> [Char] -> Bool
    result (Proposition unOp prop) list poss = checkUnOP (getPropositionValue prop list poss) unOp
    result (Expression unOp v1 binOp v2) list poss | binOp == "&" = checkUnOP ((result v1 list poss) && (result v2 list poss)) unOp
                                                   | binOp == "|" = checkUnOP ((result v1 list poss) || (result v2 list poss)) unOp
                                                   | binOp == "*" = checkUnOP ((not (result v1 list poss)) || (result v2 list poss)) unOp
                                                   | otherwise = checkUnOP (((result v1 list poss) && (result v2 list poss)) || (( not (result v1 list poss)) && ( not (result v2 list poss)))) unOp

    {-
    Transforma um booleano em uma string representativa "0" ou "1".
    Ex:
    resToString True = "1"
    resToString False = "0"
    -}
    resToString :: Bool -> String
    resToString True = "1"
    resToString False = "0"


    --                                   Funlões para a montagem da tabela

    {-
    Monta a primeira linha da tabela.
    Ex:
    firstLine "abc" "(a & (b | c))" = a b c | (a & (b | c))
    -}
    firstLine :: [Char] -> [Char] -> [Char]
    firstLine "" exp = "| " ++ exp
    firstLine (x:xs) exp = [x] ++ " " ++ (firstLine xs exp)  

    {-
    Monta cada linha da tabela.
    Ex:
    otherLine "01" "01" "ab" "(a & b)" = "0 1 | 0"
    otherLine "11" "11" "ab" "(a & b)" = "1 1 | 1"
    -}
    otherLine :: [Char] -> [Char] -> [Char] -> Literal -> [Char]
    otherLine "" valor prop  lit = "| " ++ resToString(result lit prop valor) 
    otherLine (x:xs) valor prop lit = [x] ++ " " ++ (otherLine xs valor prop lit)

    {-
    Junta todas linhas da tabela.
    Ex:
    tailTable ["0","1"], "a" "(a)" =
    "
    0 | 0
    1 | 1
    "
    -}
    tailTable :: [[Char]] -> [Char] -> Literal ->[Char]
    tailTable [] prop lit = ""
    tailTable (x:xs) prop lit = otherLine x x prop lit ++ "\n" ++ tailTable xs prop lit

    {-
    Junta a cauda da tabela com a primeira linha.
    Ex:
    table ["0","1"], "a", "(-a)" = 
    "
    a | ~a
    0 | 1
    1 | 0
    "
    -}
    table :: [[Char]] -> [Char] -> Literal -> [Char]
    table lista prop lit =  (firstLine prop (literalToString lit)) ++ "\n" ++ (tailTable lista prop lit )

    {-
    Retorna a tabela do literal.
    Ex:
    literalToString lit = (a & b)
    callTable lit =
    "
    a b | (a & b)
    0 0 | 0
    0 1 | 0
    1 0 | 0
    1 1 | 1 
    "
    -}
    callTable :: Literal -> IO()
    callTable lit = do
        putStrLn $ table (callGetPossibilitiesTable (length(getPropositions (literalToString lit)))) (getPropositions (literalToString lit)) (lit)
        putStrLn "Deseja criar uma nova tabela ou voltar ao menu?"
        putStrLn "0 - Voltar ao menu!"
        putStrLn "1 - Criar nova!"
        repeatTruthTable


    --                              ##   MAIN   ##

    {-
    Funcao principal.
    Da as boas vindas e chama o menu de opcoes.
    -}
    execTruthTable :: IO()
    execTruthTable = do
        putStrLn "\nBem vindo ao gerador de Tabela Verdade!"
        printOptions
    

    {-
    Mostra as opcoes para o usuario.
    Chama a funcao de escolha de opcoes.
    -}
    printOptions :: IO()
    printOptions = do
        putStrLn "\nEscolha sua opcao! (0, 1, 2 ou 3)"
        putStrLn "0. Retornar ao Menu."
        putStrLn "1. Qual o objetivo do gerador de Tabela Verdade?"
        putStrLn "2. Tutorial como funciona o literal (util para criar a expressao)."
        putStrLn "3. Gerar a tabela verdade de uma expressao."
        escolherOpcoes

    {-
    Menu de opcoes:
    0 - Retorna ao menu principal
    1 - Chama a funcao que mostra uma explicacao sobre a tabela verdade.
    2 - Chama a funcao que explica como montar um literal.
    3 - Chama a funcao auxiliar para montar o literal.
    -}
    escolherOpcoes :: IO()
    escolherOpcoes = do
        opcao <- readLn :: IO Int
        if (opcao == 0)
        then
            putStrLn "\nVolte sempre!\n"
        else if (opcao == 1)
        then
            explicaTable
        else if (opcao == 2)
        then
            explicaLit
        else
            auxExe

    {-
    Recebe o literal do usuario e chama a funcao de montagem da tabela
    para o literal montado pelo usuario.
    -}
    auxExe :: IO()
    auxExe = do
        putStrLn "Primeiro digite o literal que deseja-se aplicar a avaliacao"
        lit <- Lit._receiveInput
        putStrLn "\nEssa eh a tabela verdade da expressao:"
        callTable lit

    {-
    Escolha do usuario:
    0 - Voltar ao menu principal
    1 - Montar outra tabela
    -}
    repeatTruthTable :: IO()
    repeatTruthTable = do
        op <- readLn :: IO Int
        if (op == 0)
        then
            putStrLn "\nVolte sempre!\n"
        else
            printOptions

    {-
    Explicacao sobre a tabela verdade.
    -}
    explicaTable :: IO()
    explicaTable = do
        putStrLn "--------------------TABELA VERDADE------------------------"
        putStrLn "\nTabela verdade eh um dispositivo utilizado no estudo da logica."
        putStrLn "Com o uso desta tabela eh possível definir o valor logico de uma expressao,"
        putStrLn "ou seja, saber quando uma sentenca eh verdadeira ou falsa.\n"
        putStrLn "Quanto mais variaveis fizerem parte da expressao, maior o numero de linhas, que eh obtido por (2 ^ n),"
        putStrLn "onde n eh o numero de variaveis."
        putStrLn "Eh simples fazer manualmente uma tabela com duas variaveis, e ate tres."
        putStrLn "Mas quanto mais adicionamos variaveis a expressao, o numero de linhas fica exponencialmente maior.\n"
        putStrLn "O objetivo dessa parte do programa eh justamente facilitar a construcao da tabela de qualquer expressao,"
        putStrLn "Nao importando o numero de variaveis.\n"
        putStrLn "-----------------------------------------------------------"
        printOptions

    {-
    Explicacao sobre literal.
    -}
    explicaLit :: IO()
    explicaLit = do
        putStrLn "---------------------LITERAL TUTORIAL----------------------"
        putStrLn "\nUm literal pode ser uma expressao ou uma proposicao.\nUma expressao eh formada por:"
        putStrLn "- Um operador unario (~, ) para dizer se o valor da expressao eh negado;"
        putStrLn "- Um valor A: esse valor pode ser uma outra expressao ou uma simples proposicao;"
        putStrLn "- Um operador binario (&,|,*,#), que serve para dizer se a expressao eh uma conjuncao, uma disjuncao, uma implicacao ou uma bi-implicacao;"
        putStrLn "- Um valor B: esse valor pode ser uma outra expressao ou uma simples proposicao;\n"     
        putStrLn "Uma proposicao eh formada por:"
        putStrLn "- Um operador unario (~, ) para dizer se o valor da proposicao eh negado;"
        putStrLn "- Um valor: esse valor eh obrigatoriamente uma string, que no caso, vai ser a letra que representa a proposicao."
        putStrLn "Agora que ja sabemos como eh formado um literal, podemos criar um a seguir:\n"
        putStrLn "-----------------------------------------------------------"
        printOptions

File: prolog/truth_table_generator.pl
% :- use_module(truth_table_generator,[]).
:- [literal].
% :- [tty].


truth_table_generator() :-
    writeln(" --------------GERADOR DE TABELA VERDADE-------------------"),
    options_menu().


options_menu() :-
    writeln("\n Escolha sua opcao! (0, 1, 2 ou 3)"),
    writeln(" 0. Retornar ao Menu."), 
    writeln(" 1. Qual o objetivo do gerador de Tabela Verdade?"),
    writeln(" 2. Tutorial como funciona o literal (util para criar a expressao)."),
    writeln(" 3. Gerar a tabela verdade de uma expressao.\n"),
    str_input(Op),
    string_to_atom(Op,Option),
    choose_option(Option).

choose_option(Op) :-
    (Op == '0'),
    writeln("\n Volte sempre!\n");
    (Op == '1'),
    explain_table();
    (Op == '2'),
    explain_literal();
    (Op == '3'),
    create_table;
    writeln("\n Opcao invalida!\n"),
    choose_option(Op).

explain_table() :-
    writeln(" --------------------TABELA VERDADE------------------------"),
    writeln("\n Tabela verdade eh um dispositivo utilizado no estudo da logica."),
    writeln(" Com o uso desta tabela eh possível definir o valor logico de uma expressao,"),
    writeln(" ou seja, saber quando uma sentenca eh verdadeira ou falsa.\n"),
    writeln(" Quanto mais variaveis fizerem parte da expressao, maior o numero de linhas, que eh obtido por (2 ^ n),"),
    writeln(" onde n eh o numero de variaveis."),
    writeln(" Eh simples fazer manualmente uma tabela com duas variaveis, e ate tres."),
    writeln(" Mas quanto mais adicionamos variaveis a expressao, o numero de linhas fica exponencialmente maior.\n"),
    writeln(" O objetivo dessa parte do programa eh justamente facilitar a construcao da tabela de qualquer expressao,"),
    writeln(" Nao importando o numero de variaveis.\n"),
    writeln(" -----------------------------------------------------------"),
    options_menu.

explain_literal() :-
    writeln(" ---------------------LITERAL TUTORIAL----------------------"),
    writeln("\n Um literal pode ser uma expressao ou uma proposicao.\nUma expressao eh formada por:"),
    writeln(" - Um operador unario (~, ) para dizer se o valor da expressao eh negado;"),
    writeln(" - Um valor A: esse valor pode ser uma outra expressao ou uma simples proposicao;"),
    writeln(" - Um operador binario (&,|,*,#), que serve para dizer se a expressao eh uma conjuncao, uma disjuncao, uma implicacao ou uma bi-implicacao;"),
    writeln(" - Um valor B: esse valor pode ser uma outra expressao ou uma simples proposicao;\n"), 
    writeln(" Uma proposicao eh formada por:"),
    writeln(" - Um operador unario (~, ) para dizer se o valor da proposicao eh negado;"),
    writeln(" - Um valor: esse valor eh obrigatoriamente uma string, que no caso, vai ser a letra que representa a proposicao."),
    writeln(" Agora que ja sabemos como eh formado um literal, podemos criar um a seguir:\n"),
    writeln(" -----------------------------------------------------------"),
    options_menu.

repeat_options() :-
    writeln(" -----------------------------------------------------------\n"),
    
    writeln(" Escolha sua opcao! (0, 1, 2 ou 3)"),
    writeln(" 0. Voltar ao menu!"),
    writeln(" 1. Qual o objetivo do gerador de Tabela Verdade?"),
    writeln(" 2. Tutorial como funciona o literal (util para criar a expressao)."),
    writeln(" 3. Criar uma nova tabela!"),
    str_input(Op),
    string_to_atom(Op,Option),
    repeat(Option).

repeat(Op) :-
    (Op == '0'),
    writeln("\n Volte sempre!\n");
    (Op == '1'),
    explain_table();
    (Op == '2'),
    explain_literal();
    (Op == '3'),
    create_table();
    writeln("\n Opcao invalida!\n"),
    repeat(Op).


create_table() :-
    writeln("\n -----------------------------------------------------------"),
    writeln("\n Primeiro monte o literal (expressao) que sera usado para obtencao da tabela\n"),
    verifyEntryAndCreatesANewLiteral(Literal),

    writeln("\n Essa eh a tabela verdade da expressao:\n"),
    writeln(" ---------------------TABELA VERDADE------------------------\n"),
    get_table(Literal,Table),
    writeln(Table),
    repeat_options().

negate(Value,Return) :-
    Value, Return = false;
    Return = true.

and(A,B,Return) :-
    A,B,Return = true; Return = false.

or(A,B,Return) :-
    (A;B),Return = true; Return = false.

implication(A,B,Return) :-
    (\+A;B),Return = true; Return = false.

xor(A,B,Return) :-
    (A,B;\+A,\+B),Return = true; Return = false.


is_proposition(Atom,Return) :-
    Ignore = ["(",")","&","|","*","#","-",">","<"," ","~"],
    atom_string(Atom, String),
    
    (member(String,Ignore)) -> Return = false; Return = true.
    
remove_repetitions([],_,[]).
remove_repetitions([H|T],Aux,Return) :-
    (member(H,Aux)) ->
    remove_repetitions(T,Aux,R1),
    Return = R1;
    append([H],Aux,Aux2),
    remove_repetitions(T,Aux2,R2),
    append([H],R2,R3),
    Return = R3.

get_propositions_aux([],Return) :- Return = [].
get_propositions_aux([H|T],Return) :-
    is_proposition(H,Proposicao),
    (Proposicao) -> get_propositions_aux(T,R),
    append([H],R,R2),
    Return = R2;
    get_propositions_aux(T,R),
    Return = R.

get_propositions(Expression,Return) :-
    string_chars(Expression,List),
    get_propositions_aux(List,R),
    remove_repetitions(R,[],R2),
    Return = R2.

dec_bin(0,'0').
dec_bin(1,'1').
dec_bin(N,B) :-
    N > 1,
    X is N mod 2,
    Y is N//2,
    dec_bin(Y,B1),
    atom_concat(B1, X, B).

len(Atom,Return) :-
    atom_chars(Atom,List),
    length(List,Len),
    Return = Len.

complete_bin(Bin,Size,Return) :-
    len(Bin,Len),
    (Len =:= Size),
    Return = Bin;
    atom_concat('0',Bin,B2),
    complete_bin(B2,Size,R2),
    Return = R2.

get_possibilities_aux(0,Fix,Return) :- 
    complete_bin('0',Fix,Bin),
    Return = [Bin].
get_possibilities_aux(Size,Fix,Return) :-
    NewSize is Size - 1,
    dec_bin(Size,R),
    get_possibilities_aux(NewSize,Fix,R2),
    complete_bin(R,Fix,B2),
    append(R2,[B2],R3),
    Return = R3.

get_possibilities(Expression,Return) :-
    get_propositions(Expression,Prop),
    length(Prop,Number),
    Lines is (2 ^ Number) - 1,
    get_possibilities_aux(Lines,Number,R),
    Return = R.





get_proposition_position(Prop,[H|T],Return) :-
    (Prop == H) -> Return is 0;
    get_proposition_position(Prop,T,R2),
    R3 is 1 + R2,
    Return = R3.
    
get_proposition_value(Prop,List,Poss,Return) :-
    get_proposition_position(Prop,List,Position),
    nth0(Position,Poss,Value),
    Return = Value.

    

check_unary_operator(Value,Operator,Return) :-
    (Operator == '~') ->
    negate(Value,V),
    Return = V;
    Return = Value.

result(proposition(UnOp,Prop),List,Poss,Return) :- 
    get_proposition_value(Prop,List,Poss,R1),
    check_unary_operator(R1,UnOp,R2),
    Return = R2.
result(expression(UnOp,V1,BinOp,V2),List,Poss,Return) :-
    (BinOp == '&') ->
    result(V1,List,Poss,A),
    result(V2,List,Poss,B),
    and(A,B,AB),
    check_unary_operator(AB,UnOp,R1),
    Return = R1;
    (BinOp == '|') ->
    result(V1,List,Poss,C),
    result(V2,List,Poss,D),
    or(C,D,CD),
    check_unary_operator(CD,UnOp,R2),
    Return = R2;
    (BinOp == '*') ->
    result(V1,List,Poss,E),
    result(V2,List,Poss,F),
    implication(E,F,EF),
    check_unary_operator(EF,UnOp,R3),
    Return = R3;
    (BinOp == '#') ->
    result(V1,List,Poss,G),
    result(V2,List,Poss,H),
    xor(G,H,GH),
    check_unary_operator(GH,UnOp,R4),
    Return = R4.


rest_to_string(true, Return) :- Return = "1".
rest_to_string(false,Return) :- Return = "0".

boolean_list([],[]).
boolean_list([H|T],Return) :-
    (H == '1') ->
    A = [true],
    boolean_list(T,R1),
    append(A,R1,R2),
    Return = R2;
    B = [false],
    boolean_list(T,R1),
    append(B,R1,R2),
    Return = R2.

first_line([],Exp,Return) :-
    concat("| ",Exp,R), Return = R.
first_line([H|T],Exp,Return) :-
    concat(H," ",R),
    first_line(T,Exp,R2),
    concat(R,R2,R3),
    Return = R3.

get_line([],Value,Prop,Literal,Return) :-
    result(Literal,Prop,Value,R1),
    rest_to_string(R1,R2),
    concat(" | ",R2,R3),
    Return = R3.
get_line([H|T],Value,Prop,Literal,Return) :-
    concat(" ",H,R1),
    get_line(T,Value,Prop,Literal,R2),
    concat(R1,R2,R3),
    Return = R3.

get_every_line([], _, _, "").
get_every_line([H|T],Prop,Literal,Return) :-
    atom_chars(H,H2),
    boolean_list(H2,H3),
    get_line(H2,H3,Prop,Literal,R1),
    concat(R1,"\n",R2),
    get_every_line(T,Prop,Literal,R3),
    concat(R2,R3,R4),
    Return = R4.

mount_table(Interpretations,Propositions,Literal,Return) :-
    literalsToString(Literal,Expression),
    first_line(Propositions,Expression,First),
    concat(" ",First,F1),
    concat(F1,"\n",F2),
    get_every_line(Interpretations,Propositions,Literal,R1),
    concat(F2,R1,R2),
    Return = R2.

get_table(Literal,Return) :-
    literalsToString(Literal,Expression),
    get_possibilities(Expression,Interpretations),
    get_propositions(Expression,Propositions),
    mount_table(Interpretations,Propositions,Literal,R1),
    Return = R1.